### DISCUSSION TITLE
Stream

### DISCUSSION THREAD

**Discussion by showbit01** [MatthiasValvekens/pyHanko#437] [Q&A]
Hi,
I have signed the document and in the signature panel it's showing document has been altered or corrupted I have used the valid certificate how it could be ?
Any idea ?

**Comment by MatthiasValvekens** [comment]
There are many ways to screw up a signature ;). You won't get any actionable feedback if you don't share your code and explain what you are trying to achieve, at the very least.

**Comment by showbit01** [comment]
Apologize for the confusion.


1.I am writing field annotation to the input pdf stream and taking the updated stream as stream.pkl for signing:

Box = [(50,0,150,40),(500,0,350,80),(500,200,350,80)]


with open(input_pdf_path, 'rb') as inf:

    w = IncrementalPdfFileWriter(inf, strict=False)


    for i in range(3):

                Field_Name='Signature_2'+str(i)
    
                append_signature_field(Flag,
                    w,
                    SigFieldSpec(
                        sig_field_name=Field_Name,
                        field_mdp_spec=fields.FieldMDPSpec(fields.FieldMDPAction.ALL),

                        doc_mdp_update_value=fields.MDPPerm.FILL_FORMS,

                        # empty_field_appearance=True,
                        on_page=i,
                 
                        box=Box[k]

                    )
                )


    buffer = io.BytesIO()

    w.write(buffer)

    x = byte_hash(buffer)
    print(x)

    with open('stream.pkl', 'wb') as file:
        pickle.dump(buffer, file)

2.After getting updated steam.pkl, i am signing pdf from this stream.

Box = [(50,0,150,40),(500,0,350,80),(500,200,350,80)]
with open('stream.pkl','rb') as file:
    w= IncrementalPdfFileWriter(inf,strict=False)

    timestamp = datetime.now()
    md_algorithm = 'sha256'

    for i in range(3):
        timestamp = datetime.now()

        data = base64.b64decode(coded_string)

        if i>0:
            w = IncrementalPdfFileWriter(output, strict=True)


      cms_writer = cms_embedder.PdfCMSEmbedder(Flag).write_cms(


            field_name=Field_Name, writer=w

        )
        sig_field_ref = next(cms_writer)


    # #assert sig_field_ref.get_object()['/T'] == 'Signature'

        appearance=cms_embedder.SigAppearanceSetup(style= stamp.TextStampStyle(
                            stamp_text=' ',
                            # text_box_style=text.TextBoxStyle(
                            #     font=opentype.GlyphAccumulatorFactory('NotoSans-Regular.ttf')
                            # ),
                            background=images.PdfImage(img_path, opacity=1), border_width=0,
                            background_layout=layout.SimpleBoxLayoutRule(
                                x_align=AxisAlignment.ALIGN_MID,
                                y_align=AxisAlignment.ALIGN_MIN,
                                # margins=Margins(left=10, right=10, top=5, bottom=5),
                                )),
                            timestamp=timestamp,
                            name='',
                            )
        sig_obj = signers.SignatureObject(timestamp=timestamp, bytes_reserved=20440)



        cms_writer.send(
            cms_embedder.SigObjSetup(
                sig_placeholder=sig_obj,
                mdp_setup=cms_embedder.SigMDPSetup(
                    md_algorithm=md_algorithm,
                    certify=True,                
                
                    docmdp_perms=fields.MDPPerm.FILL_FORMS,
                    ),

                    appearance_setup=appearance,
            )
        )

        prep_digest, output = cms_writer.send(
            cms_embedder.SigIOSetup(md_algorithm=md_algorithm, in_place=True)
        )

        hashb64 = base64.b64encode(prep_digest.document_digest).decode()

        der_bytes = data
        # if pem.detect(der_bytes):
        #     print(11)
        #     type_name, headers, der_bytes = pem.unarmor(der_bytes)
        # cert = x509.Certificate.load(der_bytes)

        sig_contents = cms_writer.send(der_bytes)



    with open(output_pdf_path, "wb") as outfile:
        outfile.write(output.read())


''''''''''''''''''''''''''''''''''''''''''''''

The signed pdf showing is corrupted or altered here the attached screenshot of signed pdf.


![image](https://github.com/MatthiasValvekens/pyHanko/assets/68598980/b6cb0859-7595-48ea-9b31-8bcf83ec302a)
